t/porting/cpphdrcheck.t

```diff
@@ -0,0 +1,580 @@
+# it but isn't core, and has it's own limitations
+
+my $cc = $Config{cc};
+$cc = shift if @ARGV;
+
+my $ccflags = $Config{ccflags};
+
+# we add a similar C++ -std
+$ccflags  =~ s/-std[:=]\S+//;
+
+my ($ccpp_cfg, $diag) = find_ccpp($cc);
+
+note @$diag;
+
+$ccpp_cfg
+  or skip_all("Cannot find a C++ compiler corresponding to $Config{cc}");
+
+note join "\n", @$diag;
+
+my $perl_headers = <<'HEADERS';
+#include "EXTERN.h"
+#include "perl.h"
+#include "XSUB.h"
+
+HEADERS
+
+$ccflags .= " " . join " ", map { "-I$_" }
+  File::Spec->catdir($cwd, ".."),
+  # win32 has special config.h handling during the build
+  File::Spec->catdir($cwd, "..", "lib", "CORE");
+
+for my $std ("base", sort keys $ccpp_cfg->{stdargs}->%*) {
+    my $code = get_source("cpp$std");
+    $code =~ s(^//PERLHEADERS$)($perl_headers)m
+      or die "Couldn't insert headers in cpp$std";
+
+    my %std_opt = $std eq "base" ? () : ( std => $std );
+    ok_compile_only({ code => \$code, %std_opt, opts => $ccflags }, $ccpp_cfg, "test std $std")
+      or diag "Code: $code";
+}
+
+done_testing();
+
+sub shellquote (@words) {
+    state $esc = $^O eq "MSWin32" ? qr/["]/ : qr/["\\]/;
+    state $need_esc = $^O eq "MSWin32" ? qr/[ "]/ : qr/[ "\\]/;
+
+    for my $word (@words) {
+        if ($word =~ $need_esc) {
+            $word =~ s/($esc)/\\$1/g;
+            $word = qq("$word");
+        }
+    }
+
+    return "@words";
+}
+
+sub find_ccpp ($cc) {
+    my $ccpp;
+    my $cfg;
+    my $exe = $Config{_exe};
+    my @pre = shellwords($cc);
+    my @post;
+    my @diag;
+
+    # FIXME: handle building with a C++ compiler eg cc=g++
+
+    # $Config{cc} is meant to be the name of the C compiler, but some people
+    # supply switches too (which belong in ccflags, ldflags and/or lddlflags)
+    #
+    # Strip anything at the end starting with "-", this don't catch all possible
+    # such options (an option may have a separate value without "-") but
+    # once this goes into production we'll just (mostly) harmlessly skip
+    # such configurations.
+    #
+    # cc may also include a wrapper like ccache or env, which we leave in @pre
+    # here.
+    while (@pre && $pre[-1] =~ /^-/) {
+        unshift @post, pop @pre;
+    }
+
+    my $ccarg = pop @pre;
+    unless ($ccarg) {
+        push @diag, qq(Nothing left after stripping arguments from "$cc"\n);
+        return (undef, \@diag);
+    }
+
+    # gcc
+    if (($ccpp = $ccarg) =~ s/\bgcc((?:-\d+)?(?:\Q$exe\E)?)$/g++$1/aa
+        && ($cfg = check_cpp_compiler(shellquote(@pre, $ccpp, @post), "gcc,unix", \@diag))) {
+        return ( $cfg, \@diag );
+    }
+    # clang
+    elsif (($ccpp = $ccarg) =~ s/\bclang((?:-\d+)?(?:\Q$exe\E)?)$/clang++$1/aa
+           && ($cfg = check_cpp_compiler(shellquote(@pre, $ccpp, @post), "clang,unix", \@diag))) {
+        return ( $cfg, \@diag );
+    }
+    # msvc
+    # may need work if we ever support clang-cl
+    elsif ($ccarg =~ m!([\\/]|^)\bcl(?:(?:\Q$exe\E)?)?$!i
```

Comment:
This `\b` is technically redundant here.
Why the doulbe `(?: )` around `$exe`?

Reply:
IIRC I was thinking of handling clang-cl here and didn't clean up after myself.